# Exercise-2
Add the specified code for each code cell, running the cells _in order_.

Open the included **`data/words_small.txt`** file (containing a "small" list of English words) and `read()` it into a single string variable **`words_str`**. Then print the first 100 _characters_ of the `words_str` variable (use a slice operator!)

In [5]:
file = open('data/words_small.txt','r')
words_str = file.read()
words_str[:100]

'abbreviate\nabbreviation\nabdomen\nable\naboard\nabroad\nabsent\nabsolute\nabuse\nacademic\nacademy\naccelerate'

**Using a loop**, count the number of times the newline character `\n` appears in the `words_str`. This should tell you the number of words in the file (plus or minus one). You can check your work with the built-in `count()` string method (but use the loop for practice!)

In [8]:
line_n = 0
for word in words_str:
    if word == '\n':
        line_n = line_n+1
print(line_n)

1520


Create a variable **`letter_counts`** that is a dictionary whose keys are characters (letters) and whosevalues are the number of times that character appears in `words_str`. Print the dictionary after you make it.
- To do this, start with an empty dictionary, then loop through the letters one at a time. At each letter, increase the value associated with that letter key by one.
- Watch out for if the letter is not in the dictionary yet!
- Also _delete_ the entry for the newline `\n` character

In [11]:
letter_counts = {}
for letter in words_str:
    letter_counts[letter] = words_str.count(letter)
letter_counts.pop('\n')
letter_counts

{'a': 703,
 'b': 188,
 'c': 328,
 'd': 314,
 'e': 1118,
 'f': 155,
 'g': 213,
 'h': 295,
 'i': 515,
 'j': 16,
 'k': 126,
 'l': 457,
 'm': 223,
 'n': 550,
 'o': 598,
 'p': 224,
 'q': 9,
 'r': 642,
 's': 504,
 't': 604,
 'u': 279,
 'v': 85,
 'w': 161,
 'x': 23,
 'y': 189,
 'z': 15}

Print out all the count in the `letter_counts` dictionary in the format `"A appears 703 times"`, each on a new line, in alphabetical order. _Hint_: iterate through the sorted keys!

In [15]:
for letter,count in sorted(letter_counts.items()):
    print(letter.upper(), 'appears', count , 'times')

A appears 703 times
B appears 188 times
C appears 328 times
D appears 314 times
E appears 1118 times
F appears 155 times
G appears 213 times
H appears 295 times
I appears 515 times
J appears 16 times
K appears 126 times
L appears 457 times
M appears 223 times
N appears 550 times
O appears 598 times
P appears 224 times
Q appears 9 times
R appears 642 times
S appears 504 times
T appears 604 times
U appears 279 times
V appears 85 times
W appears 161 times
X appears 23 times
Y appears 189 times
Z appears 15 times


Print which letter from the `letter_counts` dictionary has a **value** of `598`. _Hint:_ use a loop to look through all the _items_ until you find the right one, then print that!

In [20]:
for key in letter_counts.keys():
    if letter_counts[key] == 598:
        print(key.upper())


O


Create a variable **`count_to_letter_dict`** that is a dictionary whose **keys** are the counts and whose **values** are the letters that have that count. This is a _reverse lookup_ dictionary. Hint: loop through items in `letter_counts` and add an item to your new dictionary for each element. You can assume no duplicated counts.

In [21]:
count_to_letter_dict = {}
for letter,counts in letter_counts.items():
    count_to_letter_dict[counts] = letter
count_to_letter_dict

{9: 'q',
 15: 'z',
 16: 'j',
 23: 'x',
 85: 'v',
 126: 'k',
 155: 'f',
 161: 'w',
 188: 'b',
 189: 'y',
 213: 'g',
 223: 'm',
 224: 'p',
 279: 'u',
 295: 'h',
 314: 'd',
 328: 'c',
 457: 'l',
 504: 's',
 515: 'i',
 550: 'n',
 598: 'o',
 604: 't',
 642: 'r',
 703: 'a',
 1118: 'e'}

Print the **10** most common letters and their counts with the format `"A appears 703 times"`. _Hint:_ you'll need to sort the keys to `count_to_letter_dict` and take the first 10 of those.

In [27]:
sorted_ctld = sorted(count_to_letter_dict.keys(), reverse = True)
for key in sorted_ctld[:10]:
    print(count_to_letter_dict[key], 'appears', key,'times')

e appears 1118 times
a appears 703 times
r appears 642 times
t appears 604 times
o appears 598 times
n appears 550 times
i appears 515 times
s appears 504 times
l appears 457 times
c appears 328 times


Bonus: repeat your analysis on the `data/words_huge.txt` file! Are the top letters the same?